benchexec/container.py

```diff
@@ -531,29 +543,47 @@ def duplicate_mount_hierarchy(mount_base, temp_base, work_base, dir_modes):
             work_path = work_base + b"/" + str(overlay_count).encode()
             os.makedirs(temp_path, exist_ok=True)
             os.makedirs(work_path, exist_ok=True)
-            try:
-                # Previous mount in this place not needed if replaced with overlay dir.
-                libc.umount(mount_path)
-            except OSError as e:
-                logging.debug(e)
-            try:
-                make_overlay_mount(mount_path, mountpoint, temp_path, work_path)
-            except OSError as e:
-                mp = mountpoint.decode()
-                raise OSError(
-                    e.errno,
-                    f"Creating overlay mount for '{mp}' failed: {os.strerror(e.errno)}. "
-                    f"Please use other directory modes, "
-                    f"for example '--read-only-dir {shlex.quote(mp)}'.",
-                )
+            if os.path.ismount(mount_path):
+                try:
+                    # Previous mount in this place not needed if replaced with overlay dir.
+                    libc.umount(mount_path)
+                except OSError as e:
+                    logging.debug(e)
+            if use_fuse:
```

Comment:
I think here we need to check `fuse_overlay_mount_path` instead, right? Otherwise we enter this case even if `fuse-overlayfs` was not found.

And we need to make sure that `fuse_overlay_mount_path` always has a value, i.e., set it to `None` in the `not use_fuse` case above (instead of setting `use_fuse = False`, which is redundant).

Reply:
I think we need to check both `use_fuse` and `fuse_overlay_mount_path` here. Consider this situation: we always use kernel overlayfs, so `use_fuse` is false. After we first fall back to fuse-overlayfs, `fuse_overlay_mount_path` is no longer `None`, and every subsequent overlay must use fuse-overlayfs. This seems to violate the principle of using fuse-overlayfs only when necessary.

Reply:
Indeed, I overlooked this.